Objective C Storyboard swizzling

mpdifran opened the pull request:
DO NOT MERGE: work in progress.

EDIT: It is now ready for merge once we update TravisCI.

mpdifran commented on Sources/Storyboard+Swizzling.m:
```diff
@@ -0,0 +1,67 @@
+    if (self == [Storyboard class]) {
+        
+        // Instantiate SwinjectStoryboard if UI/NSStoryboard is trying to be instantiated.
+        if [SwinjectStoryboard isCreatingStoryboardReference] {
```
Need to `#import "SwinjectStoryboard-Swift.h"` in order to access this Swift class method, but for some reason the header is not automatically generated.

yoichitgy replied:
@mpdifran, I'll reply shortly.

yoichitgy replied:
Hi @mpdifran. I had misunderstanding. Usually a bridging header is required to access objc classes or methods from Swift. If you need to expose Swift class, you can just add `@objc` annotation. If the problem is that `isCreatingStoryboardReference` is defined as `static var`, you can change it to `class var`.

Am I missing the point of your comment?

mpdifran replied:
You're right, the bridging header is for accessing ObjC classes in Swift, but in this case, we're accessing a Swift class in ObjC. From what I've read online, there should be an automatically generated header named `SwinjectStoryboard-Swift.h` that we can import in our ObjC files. 

But maybe adding the `@objc` to the class and changing the `static var` to a `class var` will work. I'll give that a shot.

mpdifran commented:
Updated to use `@objc`, but it's still having issues finding the SwinjectStoryboard Swift class:

    /Users/mpdifran/Sandbox/SwinjectStoryboard/Sources/Storyboard+Swizzling.m:56:14: error: use of undeclared identifier 'SwinjectStoryboard'
        if ([SwinjectStoryboard isCreatingStoryboardReference]) {
             ^

yoichitgy commented:
@mpdifran, I reproduced the issue. As you said, we needed to import `SwinjectStoryboard-Swift.h` which is specified in the project as `Objective-C Generated Interface Header Name`. I checked some stackoviewflow answers, but still I couldn't find a solution. I keep solving this problem.

yoichitgy commented:
I followed the section below in the Apple's official document, but still `SwinjectStoryboard/SwinjectStoryboard-Swift.h` cannot be imported...

- "Importing Code from Within the Same Framework Target"
  - "Importing Swift into Objective-C"

https://developer.apple.com/library/content/documentation/Swift/Conceptual/BuildingCocoaApps/MixandMatch.html

mpdifran commented on Sources/Storyboard+Swizzling.m:
```diff
@@ -0,0 +1,71 @@
+
+#import "Storyboard+Swizzling.h"
+#import <objc/runtime.h>
+#import <SwinjectStoryboard/SwinjectStoryboard-Swift.h>
```
This seems to be the correct way to import the header.

mpdifran commented:
I'm still getting one warning on `public override class func initialize()` in `SwinjectStoryboard.swift`. Any ideas for how to solve that?

yoichitgy commented:
@mpdifran, I'm sorry I took time to respond.

Do you have any change since the last week? I will fetch your commit to check the warning on `public override class func initialize()`. Thanks😃

mpdifran commented:
Nope, haven't taken a look at it since.

yoichitgy commented:
@mpdifran Do you no longer have the error to import the header file? I still have error here in `Storyboard+Swizzling.m`.

```
#import <SwinjectStoryboard/SwinjectStoryboard-Swift.h>
```

mpdifran commented:
Ah looks like I hadn't pushed up my recent changes. I just did a force push, try pulling down the branch again.

Also, correct, I am able to build the project fine; I don't see the header error.

mpdifran commented on Tests/Container+SwinjectStoryboardSpec.swift:
```diff
@@ -26,7 +26,7 @@ class Container_SwinjectStoryboardSpec: QuickSpec {
                     "[\n"
                     + "    { Service: \(controllerType), Storyboard: SwinjectStoryboardTests.AnimalViewController, "
-                    + "Factory: ((Resolver, \(controllerType))) -> \(controllerType), ObjectScope: graph, InitCompleted: Specified }\n"
+                    + "Factory: (Resolver, \(controllerType)) -> \(controllerType), ObjectScope: graph, InitCompleted: Specified }\n"
```
Not sure why, but this test was failing, and I needed to make this change in order to make it pass.

yoichitgy replied:
It's difference between Swift 3.0 and 3.1. Swift 3.1 fixed the doubly printed tuple: `((a, b))`
https://github.com/Swinject/Swinject/pull/252

mpdifran replied:
Ah ok cool.

yoichitgy commented:
Thanks @mpdifran👍 I confirmed the build passed with the swift header imported. I'm going to check the issue on `public override class func initialize()`.

yoichitgy commented:
I'll investigate `public override class func initialize()` in SwinjectStoryboard.swift in this weekend😉

mpdifran commented:
Just did a rebase and force push to fix the merge conflicts. @yoichitgy I'll be around this weekend if you want some help tackling this.

mpdifran commented:
I think I've fixed the last warning. Let me know if you approve of this implementation.

mpdifran commented:
Looks like the tests are failing. I'll take another look sometime soon.

mpdifran commented:
Fixed the failing test. Moved the `+ (void)load` override into a category on `SwinjectStoryboard` instead of `UIStoryboard` in order to get the proper protocol conformance.

mpdifran commented on Sources/SwinjectStoryboard.swift:
```diff
@@ -53,14 +43,30 @@ public class SwinjectStoryboard: _SwinjectStoryboardBase, SwinjectStoryboardProt
+        bundle storyboardBundleOrNil: Bundle?) -> SwinjectStoryboard {
+        return SwinjectStoryboard.create(name: name, bundle: storyboardBundleOrNil,
+                                         container: SwinjectStoryboard.defaultContainer)
+    }
```
This needed to be done because ObjC doesn't support Swift's default parameters (`container: Resolver = SwinjectStoryboard.defaultContainer`). These two methods are functionally equivalent to the single old method, so the API hasn't actually changed.

mpdifran commented:
I think we need to update TravisCI to use Xcode 8.3.2 in order to fix the failing tests. Currently it's using Xcode 8.2, which doesn't include Swift 3.1.

mpdifran commented:
We're getting the following message on Travis:

`"Maximum number of login attempts exceeded. Please try again later."`

We might need to do this: https://github.com/Carthage/Carthage/issues/1236

yoichitgy commented:
@mpdifran, sorry I couldn't come back here. I was too busy to check this PR. I'll check the Travis configuration.

mpdifran commented:
No worries, I was able to solve it pretty quickly. Unit tests are helpful.

mpdifran commented on Sources/SwinjectStoryboard+SetUp.m:
```diff
@@ -0,0 +1,28 @@
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        if ([self conformsToProtocol:@protocol(SwinjectStoryboardProtocol)] &&
+            [[self class] respondsToSelector:@selector(setup)]) {
```
We need this line so there's no crash when the method is not implemented.

yoichitgy commented:
I'm fixing the CI error in this branch:
https://github.com/yoichitgy/SwinjectStoryboard/commits/objCSwizzle

I still get an error for `pod lib lint`. The error is reproducible in a local environment too. I'm going to investigate it tomorrow again.

mpdifran commented:
I get this when I run `pod lib lint` locally:

    'SwinjectStoryboard/SwinjectStoryboardProtocol.h' file not found

yoichitgy commented:
Yes, we should add `'Sources/*.{h,m}'` too, but a little bit complicated because `'Sources/SwinjectStoryboard.h'` should be at the end of `s.ios.source_files`...

I'm going to check the problem in this weekend again💪

yoichitgy commented:
I'm sorry, but I'll continue working on Podspec in a few days 🙇

mpdifran commented:
@yoichitgy do you think you'll have a chance to fix the pod spec soon?

yoichitgy commented:
Sorry @mpdifran for my late response for this valuable pull request😱 I was too busy for a big project as my job to work on Swinject projects, but I got knowledge about podspec a lot in the project. I will try to solve the problem within a few weeks🏃

yoichitgy commented:
@mpdifran, I resolved the following issue on `pod lib lint` in [this branch](https://github.com/yoichitgy/SwinjectStoryboard/commits/objCSwizzle).

> 'SwinjectStoryboard/SwinjectStoryboard-Swift.h' file not found

Now I got new errors:

> error: no known class method for selector 'isCreatingStoryboardReference'
> error: no known class method for selector 'createReferencedWithName:bundle:'

I'm working on the issues now.

yoichitgy commented:
The cause of the new errors is this:

https://developer.apple.com/library/content/documentation/Swift/Conceptual/BuildingCocoaApps/MixandMatch.html

> Because the generated header for a framework target is part of the framework’s public interface, only declarations marked with the public or open modifier appear in the generated header for a framework target.
>
> Swift methods and properties that are marked with the internal modifier and declared within a class that inherits from an Objective-C class are accessible to the Objective-C runtime. However, they are not accessible at compile time and do not appear in the generated header for a framework target.

> For framework targets, only declarations with the public or open modifier appear in the generated header. You can still use Swift methods and properties that are marked with the internal modifier from within the Objective-C part of your framework, as long they are declared within a class that inherits from an Objective-C class. For more information on access-level modifiers, see Access Control in The Swift Programming Language (Swift 4).

yoichitgy commented:
I solved the `pod lib lint` issues in [this branch](https://github.com/yoichitgy/SwinjectStoryboard/commits/objCSwizzle), but I want to revert the commit (e5d2043310a287dac1b30dc809d189d7f1221f6c) not to expose those methods that should not be used by SwinjectStoryboard users.

@mpdifran, do you have any idea to revert the commit and solve the problem by another way?

mpdifran commented:
Let me take a look at it when I get a chance. Was this issue introduced due to the `pod lib lint` changes? I’ve been using this branch through Carthage with no issues...

yoichitgy commented:
I'm still not sure the cause of the problem that `internal` methods are not exposed to `SwinjectStoryboard-Swift.h`. I didn't get the problem when I built SwinjectStoryboard project, but it might be reproducible if you install SwinjectStoryboard to an app project by Carthage. (I didn't try this yet.)

mpdifran commented:
I'm currently using this branch in one of my apps through Carthage, so there's no problem on the Carthage side.

yoichitgy commented:
Thanks @mpdifran👍 I'll investigate podspec to make those method `internal` in this weekend.

yoichitgy commented:
I'm trying to resolve the issue by fixing directions of Swift/ObjC references.

yoichitgy commented:
@mpdifran, I made PR #67 taking over your PR to fix the installation issue. Your PR will be automatically merged when #67 is merged.

At the end, I ended up with `performSelector` to call internal class methods defined in Swift. Not perfect but we didn't expose unnecessary methods to users of SwinjectStoryboard.

Thank you for your investigation, fix and patience😃😃😃

yoichitgy commented:
Thanks @mpdifran for merging thmye branch👍
